Add documentation, remove redundant constants to WebView (#172)
yiisoft/view: docs/use-in-web-environment.md

  will be replaced by the registered HTML code (e.g. JavaScript) targeted at the body end position.

When registering [CSS](#registering-css), [JavaScript](#registering-javascript) and [link tags](#registering-link-tags),
you must specify the position in which this tag will be rendered. Positions are implemented by public constants:

- `POSITION_HEAD` - Corresponds to the `head()` method.
- `POSITION_BEGIN` - Corresponds to the `beginBody()` method.
- `POSITION_END` - Corresponds to the `endBody()` method.
- `POSITION_READY` - Corresponds to the `endBody()` method. Located after `POSITION_END`.
- `POSITION_LOAD` - Corresponds to the `endBody()` method. Located after `POSITION_READY`.

Every Web page should have a title. You can set the title in this way:

```php
$view->setTitle('My page title');
```

Then in the view, make sure you have the following code in the `<head>` section:

```php
<title><?= \Yiisoft\Html\Html::encode($this->getTitle()) ?></title>
```

## Rendering

In addition to the `render()` and `renderFile()` methods, two `renderAjax()` and `renderAjaxString()`
methods have been added for the WEB environment to render AJAX requests.

> The `renderAjax()` method is similar to `render()` except that it will surround the view being rendered with
> the calls of `beginPage()`, `head()`, `beginBody()`, `endBody()` and `endPage()`. By doing so, the method is
> able to inject into the rendering result with `JS/CSS` scripts and files that are registered with the view.

Reviewer: ```suggestion
The `renderAjax()` method is similar to `render()` except that it will surround the view being rendered with
the calls of `beginPage()`, `head()`, `beginBody()`, `endBody()` and `endPage()`. By doing so, the method is
able to inject JavaScript, CSS, and files, registered in the view template, into the rendering result.
```

```php
$view->renderAjax('blog/posts', [
    'posts' => $posts,
]);
```

The `renderAjaxString()` method is similar to `renderAjax()`, but only accepts a ready-made string for rendering.

```php
$view->renderAjaxString('content');
```

## Registering meta tags

> Web pages usually need to generate various meta tags needed by different parties. Like page title, meta tags appear
> in the `<head>` section. In order to specify which meta tags to generate in the views, two methods are provided:

Reviewer: ```suggestion
Web pages usually need to generate various meta tags. Similar to page title, meta tags appear
in the `<head>` section. In order to specify which meta tags to add, two methods are provided:
```

```php
// Creates a meta tag from an array of attributes:
$view->registerMeta(['name' => 'keywords', 'content' => 'yii, framework, php']);

// Creates a meta tag from a `Yiisoft\Html\Tag\Meta` instance:
$view->registerMetaTag(\Yiisoft\Html\Html::meta()->name('robots')->content('noindex'));
```

> The above code will register a meta tags in the `Yiisoft\View\WebView` instance. The registered meta tag is rendered
> after the view template finishes rendering. The following HTML code will be generated and inserted at the place
> where you call `Yiisoft\View\WebView::head()` in the view template:

Reviewer: ```suggestion
The above code will register meta tags in a `Yiisoft\View\WebView` instance. The tags are added
after the view template finishes rendering. The following HTML code will be generated and inserted at the place
where you call `Yiisoft\View\WebView::head()` in the view template:
```

```html
<meta name="keywords" content="yii, framework, php">
<meta name="robots" content="noindex">
```

Note that if you call `registerMeta()` or `registerMetaTag()` multiple times, it will register multiple meta tags,

> regardless whether the meta tags are the same or not. To make sure there is only a single instance of a meta tag type,
> you can specify a key as a second parameter when calling the methods. For example, the following code registers two
> `description` meta tags. However, only the second one will be rendered.

Reviewer: ```suggestion
regardless whether the meta tags are the same or not. To make sure there is only a single instance of a meta tag of the same type,
you can specify a key as a second parameter when calling the methods. For example, the following code registers two
`description` meta tags. However, only the second one will be rendered.
```

```php
$view->registerMeta(
    ['name' => 'description', 'content' => 'This is my cool website made with Yii!'],
    'description',
);

$view->registerMetaTag(
    \Yiisoft\Html\Html::meta()->name('description')->content('This website is about funny raccoons.'),
    'description',
);
```

## Registering link tags

Like [meta tags](#registering-meta-tags), link tags are useful in many cases, such as customizing favicon, pointing to

> RSS feed or delegating OpenID to another server. You can work with link tags in the similar way as meta tags by using
> `registerLink()` or `registerLinkTag()`. For example, in a content view, you can register a link tag like follows:

Reviewer: ```suggestion
RSS feed or delegating OpenID to another server. You can work with link tags in a similar way as meta tags by using
`registerLink()` or `registerLinkTag()`. For example, in a content view, you can register a link tag like follows:
```

```php
// Creates a link tag from an array of attributes:
$view->registerLink([
    'title' => 'Live News for Yii',
    'rel' => 'alternate',
    'type' => 'application/rss+xml',
    'href' => 'https://www.yiiframework.com/rss.xml',
]);

// Creates a link tag from a `Yiisoft\Html\Tag\Link` instance:
$view->registerLinkTag(\Yiisoft\Html\Html::link('/myicon.png', [
    'rel' => 'icon',
    'type' => 'image/png',
]));
```

The code above will result in:

```html
<link title="Live News for Yii" rel="alternate" type="application/rss+xml" href="http://www.yiiframework.com/rss.xml">
<link rel="icon" type="image/png">
```

> You can use the second parameter to pass the position at which the link tag should be inserted in a page,
> default by `Yiisoft\View\WebView::POSITION_HEAD`. Similar to registering meta tags, you can specify
> a key to avoid creating duplicate link tags. Only in the `registerLink()` and `registerLinkTag()`
> methods the key is specified by the third parameter.

Reviewer: ```suggestion
You can use second parameter to specify position at which the link tag should be inserted in a page.
Default is `Yiisoft\View\WebView::POSITION_HEAD`. Similar to registering meta tags, you can specify
a key to avoid creating duplicate link tags. In the `registerLink()` and `registerLinkTag()` the key is specified as a third parameter.
```

</head>
<body>
    <script>var age = 42;</script>
</body>
```

The values of the variables will be encoded, using the `Yiisoft\Json\Json::htmlEncode()` method:

```php
$view->registerJs('data', ['username' => 'John', 'age' => 42]);
```

The code above will result in:

```html
<script>var data = {"username":"John","age":42};</script>
```

The name of variable will be used as key, preventing duplicated variable names.

## Use with an asset manager

If you are using the [yiisoft/assets](https://github.com/yiisoft/assets) package, the `Yiisoft\View\WebView`
class provides methods for batch adding CSS and JavaScript data.

```php
/**
 * @var Yiisoft\View\WebView $this
 * @var Yiisoft\Assets\AssetManager $assetManager
 */
 
$assetManager->register([
    MyAsset::class,
]);

$this->addCssFiles($assetManager->getCssFiles());
$this->addCssStrings($assetManager->getCssStrings());
$this->addJsFiles($assetManager->getJsFiles());
$this->addJsStrings($assetManager->getJsStrings());
$this->addJsVars($assetManager->getJsVars());
```

These methods process the CSS and JavaScript configuration created by the asset manager and converts it into HTML code.

## WebView Events

> The `Yiisoft\View\WebView` class trigger several events during the view rendering process. Events are classes:

Reviewer: ```suggestion
The `Yiisoft\View\WebView` class triggers several events during the view rendering process. Events are classes:
```

  Listeners of this event may set `$event->stopPropagation()` to be false to cancel the rendering process.
- `Yiisoft\View\Event\WebView\AfterRender` - triggered at the ending of rendering a view template.
  Listeners of this event can get data about the rendering results using the methods of this class.
- `Yiisoft\View\Event\WebView\PageBegin` - triggered by the call of `Yiisoft\View\WebView::beginPage()` in the view template.
- `Yiisoft\View\Event\WebView\PageEnd` - triggered by the call of `Yiisoft\View\WebView::endPage()` in the view template.
- `Yiisoft\View\Event\WebView\Head` - triggered by the call of `Yiisoft\View\WebView::head()` in the view template.
- `Yiisoft\View\Event\WebView\BodyBegin` - triggered by the call of `Yiisoft\View\WebView::beginBody()` in the view template.
- `Yiisoft\View\Event\WebView\BodyEnd` - triggered by the call of `Yiisoft\View\WebView::endBody()` in the view template.

These events are passed to the `Psr\EventDispatcher\EventDispatcherInterface` implementation,
which is specified in the constructor when the `Yiisoft\View\WebView` instance is initialized.

## Security

> When creating views that generate HTML pages, it is important that you encode and/or filter the data
> coming from end users before presenting them. Otherwise, your application may be subject to
> [cross-site scripting](http://en.wikipedia.org/wiki/Cross-site_scripting) attacks.

Reviewer: ```suggestion
When creating views that generate HTML pages, it is important that you properly encode and/or filter all the data when outputted. Otherwise, your application may be subject to
[cross-site scripting](http://en.wikipedia.org/wiki/Cross-site_scripting) attacks.
```

To display a plain text, encode it first by calling `Yiisoft\Html\Html::encode()`.
For example, the following code encodes the username before displaying it:

```php
<?php

declare(strict_types=1);

/** @var Yiisoft\View\WebView $this */
/** @var App\User\User $user */

use Yiisoft\Html\Html;
?>

<div class="username">
    <?= Html::encode($user->name) ?>
</div>
```

> For more security, to display HTML content, you can use content filtering tools such as
> [HTML Purifier](https://github.com/ezyang/htmlpurifier). Filtering tools do an excellent
> job of ensuring the security of the output, but they do not work quickly. You should
> consider caching the filtering result if your application requires high performance.

Reviewer: ```suggestion
If you need to securely display HTML content, you can use content filtering tools such as
[HTML Purifier](https://github.com/ezyang/htmlpurifier). A drawback is that it is not very performant so consider caching the result.
```
